File: mindspore/lite/tools/converter/cxx_api/converter.cc
/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "include/converter.h"
#include "include/api/data_type.h"
#include "tools/converter/cxx_api/converter_para.h"
#include "tools/converter/converter_context.h"
#include "tools/converter/converter.h"
#include "src/common/log_adapter.h"

namespace mindspore {
namespace {
constexpr size_t kMaxSectionNum = 100;
constexpr size_t kMaxConfigNumPerSection = 1000;
}  // namespace
namespace lite {
int RunConverter(const std::shared_ptr<ConverterPara> &data_);
}
Converter::Converter(converter::FmkType fmk_type, const std::vector<char> &model_file,
                     const std::vector<char> &output_file, const std::vector<char> &weight_file) {
  data_ = std::make_shared<ConverterPara>();
  if (data_ != nullptr) {
    data_->fmk_type = fmk_type;
    data_->model_file = CharToString(model_file);
    data_->output_file = CharToString(output_file);
    data_->weight_file = CharToString(weight_file);
  } else {
    MS_LOG(ERROR) << "Create ConverterPara failed";
  }
}

void Converter::SetConfigFile(const std::vector<char> &config_file) {
  if (data_ != nullptr) {
    data_->config_file = CharToString(config_file);
  }
}

std::vector<char> Converter::GetConfigFileChar() const {
  std::string cfg_file = "";
  if (data_ != nullptr) {
    cfg_file = data_->config_file;
  }
  return StringToChar(cfg_file);
}

void Converter::SetConfigInfo(const std::vector<char> &section,
                              const std::map<std::vector<char>, std::vector<char>> &config) {
  auto section_str = CharToString(section);
  auto config_str = MapVectorCharToString(config);
  if (data_ != nullptr) {
    if (data_->config_param.size() > kMaxSectionNum) {
      MS_LOG(ERROR) << "Section num " << data_->config_param.size() << "exceeds max num " << kMaxSectionNum;
      return;
    }
    if (data_->config_param.find(section_str) != data_->config_param.end()) {
      MS_LOG(WARNING) << "Section " << section_str << "already exists, "
                      << "value will be overwrite.";
    }
    if (config.size() > kMaxConfigNumPerSection) {
      MS_LOG(ERROR) << "Config num " << config.size() << " exceeds max num " << kMaxConfigNumPerSection << " in "
                    << section_str;
      return;
    }
    data_->config_param[section_str] = config_str;
  }
}

std::map<std::vector<char>, std::map<std::vector<char>, std::vector<char>>> Converter::GetConfigInfoChar() const {
  return MapMapStringToChar(data_->config_param);
}

void Converter::SetWeightFp16(bool weight_fp16) {
  if (data_ != nullptr) {
    data_->weight_fp16 = weight_fp16;
  }
}

bool Converter::GetWeightFp16() const {
  if (data_ != nullptr) {
    return data_->weight_fp16;
  } else {
    return false;
  }
}

void Converter::SetInputShape(const std::map<std::vector<char>, std::vector<int64_t>> &input_shape) {
  auto input_shape_str = MapCharToString(input_shape);
  if (data_ != nullptr) {
    for (auto &it : input_shape_str) {
      lite::ConverterInnerContext::GetInstance()->UpdateGraphInputTensorShape(it.first, it.second);
    }
    data_->input_shape = input_shape_str;
  }
}

std::map<std::vector<char>, std::vector<int64_t>> Converter::GetInputShapeChar() const {
  std::map<std::string, std::vector<int64_t>> input_shape = {};
  if (data_ != nullptr) {
    input_shape = data_->input_shape;
  }
  return MapStringToChar(input_shape);
}

void Converter::SetInputFormat(Format format) {
  if (data_ != nullptr) {
    if (format != DEFAULT_FORMAT) {
      data_->input_format = format;
    }
    data_->spec_input_format = format;
  }
}

Format Converter::GetInputFormat() const {
  if (data_ != nullptr) {
    return data_->input_format;
  } else {
    return DEFAULT_FORMAT;
  }
}

void Converter::SetInputDataType(DataType data_type) {
  if (data_ != nullptr) {
    data_->input_data_type = data_type;
  }
}

DataType Converter::GetInputDataType() {
  if (data_ != nullptr) {
    return data_->input_data_type;
  } else {
    return DataType::kTypeUnknown;
  }
}

void Converter::SetOutputDataType(DataType data_type) {
  if (data_ != nullptr) {
    data_->output_data_type = data_type;
  }
}

DataType Converter::GetOutputDataType() {
  if (data_ != nullptr) {
    return data_->output_data_type;
  } else {
    return DataType::kTypeUnknown;
  }
}

void Converter::SetExportMindIR(ModelType export_mindir) {
  if (data_ != nullptr) {
    data_->export_mindir = export_mindir;
  }
}

ModelType Converter::GetExportMindIR() const {
  if (data_ != nullptr) {
    return data_->export_mindir;
  } else {
    return kMindIR_Lite;
  }
}

void Converter::SetDecryptKey(const std::vector<char> &key) {
  if (data_ != nullptr) {
    data_->decrypt_key = CharToString(key);
  }
}

std::vector<char> Converter::GetDecryptKeyChar() const {
  std::string decrypt_key = "";
  if (data_ != nullptr) {
    decrypt_key = data_->decrypt_key;
  }
  return StringToChar(decrypt_key);
}

void Converter::SetDecryptMode(const std::vector<char> &mode) {
  if (data_ != nullptr) {
    data_->decrypt_mode = CharToString(mode);
  }
}

std::vector<char> Converter::GetDecryptModeChar() const {
  std::string decrypt_mode = "";
  if (data_ != nullptr) {
    decrypt_mode = data_->decrypt_mode;
  }
  return StringToChar(decrypt_mode);
}

void Converter::SetEnableEncryption(bool encryption) {
  if (data_ != nullptr) {
    data_->enable_encryption = encryption;
  }
}

bool Converter::GetEnableEncryption() const {
  if (data_ != nullptr) {
    return data_->enable_encryption;
  } else {
    return false;
  }
}

void Converter::SetEncryptKey(const std::vector<char> &key) {
  if (data_ != nullptr) {
    data_->encrypt_key = CharToString(key);
  }
}

std::vector<char> Converter::GetEncryptKeyChar() const {
  std::string encrypt_key = "";
  if (data_ != nullptr) {
    encrypt_key = data_->encrypt_key;
  }
  return StringToChar(encrypt_key);
}

void Converter::SetInfer(bool infer) {
  if (data_ != nullptr) {
    data_->pre_infer = infer;
  }
}

bool Converter::GetInfer() const {
  if (data_ != nullptr) {
    return data_->pre_infer;
  } else {
    return false;
  }
}

void Converter::SetTrainModel(bool train_model) {
  if (data_ != nullptr) {
    data_->train_model = train_model;
  }
}

bool Converter::GetTrainModel() const {
  if (data_ != nullptr) {
    return data_->train_model;
  } else {
    return false;
  }
}

void Converter::SetNoFusion(bool no_fusion) {
  if (data_ != nullptr) {
    data_->no_fusion = no_fusion;
  }
}

bool Converter::GetNoFusion() {
  if (data_ != nullptr) {
    return data_->no_fusion;
  } else {
    return false;
  }
}

void Converter::SetOptimizeTransformer(bool optimizeTransformer) {
  if (data_ != nullptr) {
    data_->optimize_transformer = optimizeTransformer;
  }
}

bool Converter::GetOptimizeTransformer() {
  if (data_ != nullptr) {
    return data_->optimize_transformer;
  } else {
    return false;
  }
}

void Converter::SetDevice(const std::vector<char> &device) {
  if (data_ != nullptr) {
    data_->device = CharToString(device);
  }
}

std::vector<char> Converter::GetDeviceChar() {
  std::string device = "";
  if (data_ != nullptr) {
    device = data_->device;
  }
  return StringToChar(device);
}

Status Converter::Convert() {
  if (data_ != nullptr) {
    Status ret = Status(static_cast<StatusCode>(lite::RunConverter(data_, nullptr, nullptr, false)));
    data_->decrypt_key.clear();  // clear key
    data_->encrypt_key.clear();  // clear key
    if (ret != kSuccess) {
      MS_LOG(ERROR) << "Convert model failed, ret=" << ret;
    }
    return ret;
  } else {
    return kLiteError;
  }
}

void *Converter::Convert(size_t *data_size) {
  void *model_data = nullptr;
  if (data_ != nullptr) {
    Status ret = Status(static_cast<StatusCode>(lite::RunConverter(data_, &model_data, data_size, true)));
    data_->decrypt_key.clear();  // clear key
    data_->encrypt_key.clear();  // clear key
    if (ret != kSuccess) {
      MS_LOG(ERROR) << "Convert model failed, ret=" << ret;
    }
  } else {
    MS_LOG(ERROR) << "Convert model failed, data is null.";
  }
  return model_data;
}
}  // namespace mindspore

File: mindspore/lite/tools/mslite_bench/mslite_bench/tools/converter.py
# Copyright 2023 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
""" converter for mslite """

import copy
from enum import Enum
import os
from dataclasses import  dataclass

import mindspore_lite as mslite

from mslite_bench.utils import InferLogger
from mslite_bench.common.model_info_enum import DeviceType
from mslite_bench.common.task_common_func import CommonFunc
from mslite_bench.tools.cross_framework_accuracy import CrossFrameworkAccSummary


class InputModelType(Enum):
    """ enum model type class for model to be convterted"""
    PB = mslite.FmkType.TF
    CAFFE = mslite.FmkType.CAFFE
    ONNX = mslite.FmkType.ONNX
    MINDIR = mslite.FmkType.MINDIR
    TFLITE = mslite.FmkType.TFLITE
    PTH = mslite.FmkType.PYTORCH


class MsliteModelType(Enum):
    """ enum model type in mslite"""
    MINDIR = mslite.ModelType.MINDIR
    MINDIR_LITE = mslite.ModelType.MINDIR_LITE


class MsliteDataType(Enum):
    """ enum data type for quant """
    FLOAT32 = mslite.DataType.FLOAT32
    INT8 = mslite.DataType.INT8
    UINT8 = mslite.DataType.UINT8
    UNKNOWN = mslite.DataType.UNKNOWN


class MsliteTensorFormat(Enum):
    """ enum input data format"""
    NCHW = mslite.Format.NCHW
    NHWC = mslite.Format.NHWC


@dataclass
class ConverterParams:
    """data class for converter input params"""
    input_shape: str = None
    input_data_type: str = "FLOAT32"
    output_data_type: str = "FLOAT32"
    input_format: str = "NCHW"
    weight_fp16: bool = False
    save_type: str = "MINDIR"
    decrypt_key: str = None
    decrypt_mode: str = None
    enable_encryption: bool = False
    encrypt_key: str = None
    infer: bool = False
    optimize: str = None
    device: str = DeviceType.CPU.value
    converter_output_file: str = None
    model_file: str = None
    params_file: str = None


class MsliteConverter:
    """model converter for mindspore lite"""
    @classmethod
    def convert(cls,
                args,
                logger=None,
                is_delete_ms_model=False):
        """convert third party model to mslite model"""
        if logger is None:
            logger = InferLogger(args.log_path).logger

        logger.debug('Start to convert model')
        cls.model_convert(args, logger)

        if not args.converter_is_analysis:
            return

        if cls.enum_value(args.converter_save_type, MsliteModelType) == MsliteModelType.MINDIR.value:
            output_file_name = f'{args.converter_output_file}.mindir'
        else:
            output_file_name = f'{args.converter_output_file}.ms'

        if os.path.exists(output_file_name):
            args_copy = copy.deepcopy(args)
            args_copy.cmp_model_file = args.model_file
            args_copy.model_file = output_file_name
            logger.info('Start accuracy compare procedure')
            CrossFrameworkAccSummary.accuracy_compare_func(args_copy, logger)

            if is_delete_ms_model:
                os.remove(output_file_name)

    @classmethod
    def model_convert(cls, args, logger=None):
        """model convert for mslite"""
        if logger is None:
            logger = InferLogger(args.log_path).logger
        converter = cls._init_converter(args)
        try:
            extension = os.path.splitext(args.model_file)[1][1:].upper()
            model_type = cls.enum_value(extension, InputModelType)
        except NotImplementedError as e:
            logger.error('Input model type error: %s', e)
            return

        converter.convert(model_type,
                          args.model_file,
                          args.converter_output_file,
                          weight_file=args.params_file,
                          config_file=args.converter_config_file)


    @classmethod
    def _init_converter(cls, args):
        """init mslite model converter with args"""
        converter = mslite.Converter()

        converter.input_shape = CommonFunc.get_tensor_shapes(args.converter_input_shape)
        converter.input_data_type = cls.enum_value(args.quant_input_data_type.upper(),
                                                   MsliteDataType)
        converter.output_data_type = cls.enum_value(args.quant_output_data_type.upper(),
                                                    MsliteDataType)
        converter.input_format = cls.enum_value(args.converter_input_format,
                                                MsliteTensorFormat)
        converter.weight_fp16 = args.converter_weight_fp16
        converter.save_type = cls.enum_value(args.converter_save_type, MsliteModelType)
        converter.decrypt_key = args.converter_decrypt_key
        converter.decrypt_mode = args.converter_decrypt_mode
        converter.enable_encryption = args.converter_enable_encryption
        converter.encrypt_key = args.converter_encrypt_key
        converter.infer = False
        converter.optimize = args.converter_optimize
        if args.device.lower() == DeviceType.ASCEND.value:
            converter.device = "Ascend"

        return converter

    @staticmethod
    def enum_value(enum_key, enum_class):
        """get value from key in enum class"""
        if enum_key in enum_class.__members__:
            return enum_class[enum_key].value
        raise NotImplementedError(f"{enum_key} is not in class {enum_class}")
